### PR TITLE
Issue 88: Set default values when not specified

### LINKED ISSUE #88: Set default values when not specified
The operator should use default values when they are not set in the manifest. 

The proposed default values are the following:

```
spec:
  zookeeperUri: zk-client:2181

  externalAccess:
    enabled: false

  bookkeeper:
    image:
      repository: pravega/bookkeeper
      tag: latest
      pullPolicy: Always

    replicas: 3

    storage:
      ledgerVolumeClaimTemplate:
        accessModes: [ "ReadWriteOnce" ]
        storageClassName: "standard"
        resources:
          requests:
            storage: 10Gi

      journalVolumeClaimTemplate:
        accessModes: [ "ReadWriteOnce" ]
        storageClassName: "standard"
        resources:
          requests:
            storage: 10Gi

    autoRecovery: true

  pravega:
    controllerReplicas: 1
    segmentStoreReplicas: 3

    debugLogging: false

    cacheVolumeClaimTemplate:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: "standard"
      resources:
        requests:
          storage: 20Gi

    image:
      repository: pravega/pravega
      tag: latest
      pullPolicy: Always

    tier2:
      filesystem:
        persistentVolumeClaim:
          claimName: pravega-tier2
```

This will allow users to partially fill the manifest or even leave the specs empty. The below example should be a valid manifest that would use default values:

```
apiVersion: "pravega.pravega.io/v1alpha1"
kind: "PravegaCluster"
metadata:
  name: "pravega"
```

### PR DESCRIPTION
**Pull request by adrianmo**
## Change log description
- If a `PravegaCluster` manifest is submitted with partially filled (or even with an empty spec), the operator will complete the resource specification with the default values outlined in #88. 
- Fix `PravegaCluster` spec not showing when describing a resource (e.g. `kubectl describe PravegaCluster pravega`).
- Add descriptive comments to API fields

## Purpose of the change
- Fix #88 

## How to test the change

Checkout the repository (`master` branch is fine, no need to checkout this branch).

Edit the operator image used in the `deploy/operator.yaml` file to use the `adrianmo/pravega-operator:issue-88` image, which contains the changes in this PR.

Follow the regular README file to install the dependencies and the operator.

Create and submit a CRD manifest with an empty spec.

```
apiVersion: "pravega.pravega.io/v1alpha1"
kind: "PravegaCluster"
metadata:
  name: "pravega"
```

Check that resources are created as usual.

```
$ kubectl get all -l pravega_cluster=pravega
NAME                                              READY     STATUS    RESTARTS   AGE                                                                           
pod/pravega-bookie-0                              1/1       Running   0          16m                                                                           
pod/pravega-bookie-1                              1/1       Running   0          16m                                                                           
pod/pravega-bookie-2                              1/1       Running   0          16m                                                                           
pod/pravega-pravega-controller-84f6b77c64-gjvlt   1/1       Running   0          16m                                                                           
pod/pravega-pravega-segmentstore-0                1/1       Running   1          16m                                                                           

NAME                                            TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)              AGE                                             
service/pravega-bookie-headless                 ClusterIP   None            <none>        3181/TCP             16m                                             
service/pravega-pravega-controller              ClusterIP   10.59.251.157   <none>        10080/TCP,9090/TCP   16m                                             
service/pravega-pravega-segmentstore-headless   ClusterIP   None            <none>        12345/TCP            16m                                             

NAME                                               DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE                                                            
deployment.extensions/pravega-pravega-controller   1         1         1            1           16m                                                            

NAME                                                          DESIRED   CURRENT   READY     AGE                                                                
replicaset.extensions/pravega-pravega-controller-84f6b77c64   1         1         1         16m                                                                

NAME                                         DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE                                                                  
deployment.apps/pravega-pravega-controller   1         1         1            1           16m                                                                  

NAME                                                    DESIRED   CURRENT   READY     AGE                                                                      
replicaset.apps/pravega-pravega-controller-84f6b77c64   1         1         1         16m                                                                      

NAME                                            DESIRED   CURRENT   AGE
statefulset.apps/pravega-bookie                 3         3         16m
statefulset.apps/pravega-pravega-segmentstore   1         1         16m
```

Check that the resource spec is completed with the default values.

```
$ kubectl describe PravegaCluster pravega
...
Spec:
  Bookkeeper:
    Auto Recovery:  true
    Image:
      Pull Policy:  Always
      Repository:   pravega/bookkeeper
      Tag:          latest
    Replicas:       3
    Storage:
      Journal Volume Claim Template:
        Access Modes:
          ReadWriteOnce
        Resources:
          Requests:
            Storage:  10Gi
      Ledger Volume Claim Template:
        Access Modes:
          ReadWriteOnce
        Resources:
          Requests:
            Storage:  10Gi
  External Access:
    Enabled:  false
  Pravega:
    Cache Volume Claim Template:
      Access Modes:
        ReadWriteOnce
      Resources:
        Requests:
          Storage:        20Gi
    Controller Replicas:  1
    Debug Logging:        false
    Image:
      Pull Policy:  Always
      Repository:   pravega/pravega
      Tag:          latest
    Options:
    Segment Store Replicas:  1
    Tier 2:
      Filesystem:
        Persistent Volume Claim:
          Claim Name:  pravega-tier2
  Zookeeper Uri:       zk-client:2181
```

### CODE REVIEW

**Comment by adrianmo** [comment]
I need to rebase this PR and resolve the conflicts.

**Comment by adrianmo** [comment]
@fpj @Tristan1900 the PR is now ready to review.

File: pkg/apis/pravega/v1alpha1/bookkeeper.go [hunk 1]
```diff
@@ -11,18 +11,137 @@
...
+	// Bookkeeper ledger volume
+	DefaultBookkeeperLedgerVolumeSize = "10Gi"
+
+	// DefaultBookkeeperJournalVolumeSize is the default volume size for the
+	// Bookkeeper journal volume
+	DefaultBookkeeperJournalVolumeSize = "10Gi"
+
+	// MinimumBookkeeperReplicas is the minimum number of Bookkeeper replicas
+	// accepted
+	MinimumBookkeeperReplicas = 3
```
Comment (fpj) [inline pkg/apis/pravega/v1alpha1/bookkeeper.go]: The other constants are prefixed with `Default`, why is this one not prefixed?

File: pkg/apis/pravega/v1alpha1/pravega.go [hunk 2]
```diff
@@ -11,35 +11,175 @@
...
+	// Ecs is used to configure a Dell EMC ECS system as a Tier 2 backend
+	Ecs *ECSSpec `json:"ecs,omitempty"`
+
+	// Hdfs is used to configure an HDFS system as a Tier 2 backend
+	Hdfs *HDFSSpec `json:"hdfs,omitempty"`
+}
+
+func (s *Tier2Spec) withDefaults() {
+	if s.FileSystem == nil && s.Ecs == nil && s.Hdfs == nil {
+		fs := &FileSystemSpec{
```
Comment (fpj) [inline pkg/apis/pravega/v1alpha1/pravega.go]: Naïve question: is it possible that there is no file system PV configured? If so, how does the deployment fail?

File: pkg/apis/pravega/v1alpha1/bookkeeper.go (hunk 1 above)
Comment (adrianmo) [reply pkg/apis/pravega/v1alpha1/bookkeeper.go]: In this particular case, I thought that `Minimum` better represents the purpose of this constant.

File: pkg/apis/pravega/v1alpha1/pravega.go (hunk 2 above)
Comment (adrianmo) [reply pkg/apis/pravega/v1alpha1/pravega.go]: Yes, it is possible to create a `PravegaCluster` that references a non-existent PVC. The deployment will succeed, but the Segment Store pods will fail to start and will remain in `Pending` state until the PVC is created (`pravega-tier2` by default).

Example:

```
$ kubectl get po
...
pravega-pravega-segmentstore-0                0/1     Pending             0          22s
```
A user can describe the pod to find out what the problem is:

```
$ kubectl describe po pravega-pravega-segmentstore-0
...
Events:
  Type     Reason            Age                    From               Message
  ----     ------            ----                   ----               -------
  Warning  FailedScheduling  3m10s (x4 over 3m13s)  default-scheduler  PersistentVolumeClaim is not bound: "cache-pravega-pravega-segmentstore-0" (repeated 4 times)
  Warning  FailedScheduling  7s (x12 over 3m6s)     default-scheduler  persistentvolumeclaim "pravega-tier2" not found (repeated 4 times)
```

Segment Store pods will resume their creation process as soon as the PVC is created.

```
$ kubectl create -f pvc.yaml
$ kubectl get po
...
pravega-pravega-segmentstore-0                0/1     ContainerCreating   0          4m16s
...
pravega-pravega-segmentstore-0                1/1     Running   0          5m4s
```

File: pkg/apis/pravega/v1alpha1/bookkeeper.go (hunk 1 above)
Comment (fpj) [reply pkg/apis/pravega/v1alpha1/bookkeeper.go]: Ok, this value is playing two roles then: default and minimum.

**Pull request merged**